From eef6920c30959d68cf34dfd98fdc0d220f58a3c7 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Thu, 10 Oct 2024 10:08:03 +0200
Subject: [PATCH 0895/1013] evl/net: dev: perform lockless pre-check for eBPF
 filter

Don't enter an RCU read-locked section if there is no filter installed
on entry. For this, let's perform a quick check using READ_ONCE()
before actually entering the RCU read locked section if need be.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 kernel/evl/net/dev.c | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/kernel/evl/net/dev.c b/kernel/evl/net/dev.c
index 8342fd884d53..894d4c864ff6 100644
--- a/kernel/evl/net/dev.c
+++ b/kernel/evl/net/dev.c
@@ -466,16 +466,18 @@ int evl_netdev_event(struct notifier_block *ev_block,
 enum evl_net_rx_action
 __evl_net_filter_rx(struct evl_netdev_state *est, struct sk_buff *skb)
 {
+	struct evl_net_ebpf_filter *filter = READ_ONCE(est->rx_filter);
 	enum evl_net_rx_action ret = EVL_RX_VLAN;
-	struct evl_net_ebpf_filter *filter;
 
-	rcu_read_lock();
+	if (filter) {
+		rcu_read_lock(); // Recheck under lock.
 
-	filter = rcu_dereference(est->rx_filter);
-	if (filter)
-		ret = bpf_prog_run_clear_cb(filter->prog, skb);
+		filter = rcu_dereference(est->rx_filter);
+		if (filter)
+			ret = bpf_prog_run_clear_cb(filter->prog, skb);
 
-	rcu_read_unlock();
+		rcu_read_unlock();
+	}
 
 	return ret;
 }
-- 
2.30.2

